sdk/trace/provider.go

```diff
@@ -234,6 +234,16 @@ func (p *TracerProvider) Shutdown(ctx context.Context) error {
 	return nil
 }
 
+type TracerProviderOption interface {
+	apply(*tracerProviderConfig)
+}
+
+type traceProviderOptionFunc func(*tracerProviderConfig)
+
+func (fn traceProviderOptionFunc) apply(cfg *tracerProviderConfig) {
+	fn(cfg)
+}
```

Comment:
@MrAlias @Aneurysm9

Do you accept such implementation as well? Do you think it breaks this rule?

> All user configurable options for a config must have a related unexported implementation of the Option interface and an exported configuration function that wraps this implementation.

The "must have a related unexported implementation" concerns me. Like in theory we have a related `traceProviderOptionFunc ` unexported implementation. Noone told that it has to be 1:1 relation.

IMO it avoids a lot of noise in code and the exported API is not affected.

Reply:
I'm personally a fan of using the `type fooOptionFunc func(*fooConfig)` pattern to provide a simple way to implement the `FooOption` interface with a closure where a data structure isn't required to implement the option.

```go
func WithTimestamp(t time.Time) Option {
	return optionFunc(func(c *cfg) {
		c.timestamp = t
	})
}
```

is a lot simpler than:

```go
type timestampOption time.Time

func (t timestampOption) apply(c *cfg) {
	c.timestamp = t
}

func WithTimestamp(t time.Time) Option {
	return timestampOption(t)
}
```

I think it is worth noting that this is an option in the contribution style guide, and possibly converting many of the existing options that have trivial implementations to use it.

Reply:
@Aneurysm9  Yeah. We are on the same page. I will try to create a separate PR to enhance the docs then.

Reply:
We should probably ensure @MrAlias is also on the same page.

Reply:
Sounds good to me as long as that type with the underlying function type is not exported and we are exporting the option interface we have documented so far.

Reply:
> I think it is worth noting that this is an option in the contribution style guide

I agree. I will do it tomorrow

>  possibly converting many of the existing options that have trivial implementations to use it.

This is an internal and therefore low priority. I will do it only if it will be faster for me. I want to focus on the exported API surface.

Reply:
PR: https://github.com/open-telemetry/opentelemetry-go/pull/1899 (took me more time sorry)